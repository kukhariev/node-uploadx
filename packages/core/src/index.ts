export * from './handlers';
export * from './storages';
export * from './utils';
export * from './types';
