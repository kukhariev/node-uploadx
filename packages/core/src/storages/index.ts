export * from './disk-storage';
export * from './file';
export * from './storage';
export * from './local-meta-storage';
export * from './meta-storage';
export * from './config';
