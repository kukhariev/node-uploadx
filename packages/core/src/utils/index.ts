export * from './cache';
export * from './errors';
export * from './fs';
export * from './http';
export * from './locker';
export * from './logger';
export * from './pipes';
export * from './primitives';
export * from './validator';
