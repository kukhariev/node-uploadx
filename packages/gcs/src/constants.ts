export const BUCKET_NAME = 'node-uploadx';
export const uploadAPI = `https://storage.googleapis.com/upload/storage/v1/b`;
export const storageAPI = `https://storage.googleapis.com/storage/v1/b`;
export const authScopes = ['https://www.googleapis.com/auth/devstorage.full_control'];
