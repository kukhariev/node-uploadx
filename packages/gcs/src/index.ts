export * from './gcs-storage';
export * from './gcs-meta-storage';
export * from './constants';
