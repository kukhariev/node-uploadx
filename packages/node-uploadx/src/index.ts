export * from '@uploadx/core';
export * from '@uploadx/s3';
export * from '@uploadx/gcs';
