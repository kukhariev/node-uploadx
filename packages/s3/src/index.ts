export * from './s3-storage';
export * from './s3-meta-storage';
export * from './aws-error';
