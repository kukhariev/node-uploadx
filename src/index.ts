export * from './core';
export * from './core/utils';
export * from './Tus';
export * from './Uploadx';
export * from './Multipart';
