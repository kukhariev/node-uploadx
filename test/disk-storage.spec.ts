import { join } from 'path';
import { DiskStorage, File } from '../src';
import { filename, metafile, root, storageOptions, testfile } from './fixtures';
import { PassThrough } from 'stream';

const directory = join(root, 'ds-test');
const options = { ...storageOptions, directory };

class FileWriteStream extends PassThrough {
  get bytesWritten(): number {
    return super.readableLength;
  }

  close(): void {
    return;
  }
}

let writeStream: FileWriteStream;

jest.mock('../src/utils/fs', () => ({
  ensureFile: async () => 0,
  getFiles: async () => [join(directory, filename), join(directory, metafile)],
  getWriteStream: () => writeStream,
  fsp: {
    stat: async () => ({ mtime: new Date() }),
    writeFile: async () => 0,
    readFile: async () => JSON.stringify(testfile),
    unlink: async () => null
  }
}));

describe('DiskStorage', () => {
  const storage = new DiskStorage(options);
  let mockReadable: PassThrough;

  beforeEach(async () => {
    writeStream = new FileWriteStream();
    mockReadable = new PassThrough();
    await storage.create({} as any, testfile);
  });

  it('should create file', async () => {
    const { size, status, bytesWritten } = await storage.create({} as any, testfile);
    expect(bytesWritten).toBe(0);
    expect(size).toBe(testfile.size);
    expect(status).toBe('created');
  });

  it('should update metadata', async () => {
    const file = await storage.update(filename, { metadata: { name: 'newname.mp4' } } as File);
    expect(file.metadata.name).toBe('newname.mp4');
    expect(file.metadata.mimeType).toBe('video/mp4');
  });

  it('should resume', async () => {
    const file = await storage.write({ ...testfile });
    expect(file.bytesWritten).toBe(0);
  });

  it('should write error', async () => {
    setTimeout(() => {
      mockReadable.emit('data', '12345');
      mockReadable.emit('data', '12345');
      writeStream.emit('error', 'write error');
    }, 100);
    const stream = storage.write({ ...testfile, start: 0, body: mockReadable });
    await expect(stream).rejects.toHaveProperty('statusCode', 500);
  });

  it('should close on abort', async () => {
    const close = spyOn(writeStream, 'close');
    setTimeout(() => {
      mockReadable.emit('data', '12345');
      mockReadable.emit('aborted');
      mockReadable.emit('end');
    }, 100);
    const file = await storage.write({ ...testfile, start: 0, body: mockReadable });
    expect(+file.bytesWritten).toBeNaN();
    expect(close).toBeCalled();
  });

  it('should write', async () => {
    setTimeout(() => {
      mockReadable.emit('data', '12345');
      mockReadable.emit('data', '12345');
      mockReadable.emit('end');
    }, 100);

    const file = await storage.write({ ...testfile, start: 0, body: mockReadable });
    expect(file.status).toBe('part');
    expect(file.bytesWritten).toBe(10);
  });

  it('should return files', async () => {
    const files = await storage.get(testfile.userId);
    expect(files).toEqual(expect.any(Array));
    expect(files.length).toEqual(1);
    expect(files[0]).toMatchObject({ name: filename });
  });

  it('should return file', async () => {
    const files = await storage.get(filename);
    expect(files).toEqual(expect.any(Array));
  });

  it('should delete file', async () => {
    const [deleted] = await storage.delete(filename);
    expect(deleted.name).toBe(filename);
    expect(deleted.status).toBe<File['status']>('deleted');
  });
});
