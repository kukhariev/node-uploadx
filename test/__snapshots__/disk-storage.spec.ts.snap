// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiskStorage .create() should reject on limits 1`] = `
{
  "code": "RequestEntityTooLarge",
  "message": "Request entity too large",
  "name": "ValidationError",
  "statusCode": 413,
}
`;

exports[`DiskStorage .create() should set status and bytesWritten 1`] = `
DiskFile {
  "bytesWritten": 0,
  "contentType": "video/mp4",
  "createdAt": "2022-02-02T00:00:00.000Z",
  "expiredAt": "2022-02-02T01:00:00.000Z",
  "id": "f7d13faa74e2475f-e8fed598250d10ea-7f59007b4b7cf67-120941ca7dc37b78",
  "metadata": {
    "custom": "",
    "lastModified": 1635398061454,
    "mimeType": "video/mp4",
    "name": "testfile.mp4",
    "sha1": "ZAPAntzKARqtb+j3B529GAOf3kI=",
    "size": 64,
  },
  "name": "userId/testfile.mp4",
  "originalName": "testfile.mp4",
  "size": 64,
  "status": "created",
  "userId": "userId",
}
`;

exports[`DiskStorage .list() should return all user files 1`] = `
[
  {
    "createdAt": 2022-02-02T00:00:00.000Z,
    "id": "f7d13faa74e2475f-e8fed598250d10ea-7f59007b4b7cf67-120941ca7dc37b78",
    "modifiedAt": 2022-02-02T00:00:00.000Z,
  },
]
`;
